Improve the examples for new users.

SuperDrew opened the pull request:
To get the examples working:
- [x] Move Generator to end of library module list and the Database to the start so that cabal repl starts in Database module.
- [x] Remove the square brackets generated from examples (x in example is contained in [])
- [x] Add example of writing generated sql to a file

Commit:
Move Generator to end of library module list so that cabal repl starts in Database module.

fffej commented:
The cabal changes look good, thanks!

By remove the `[]` from the examples, what'd you mean?  Do you mean that the list syntax is getting written to the file?

SuperDrew commented:
Yes, so in the example:

```
x <- sample' (arbitrary :: Gen Table)
```

x yields:

```
[CREATE TABLE UUID_660438bfadf8884fde7557f7e629edcb
(UUID_0ad776feb6f102ef9da7cb47bf234969 geography)
...
UUID_b7f5a87be09c4cadcbd21e2b1de806fd xml SPARSE)
GO]
```

This is true whether you are just showing x or writing it to a file.

Commit:
Add writing out to a file to README.
It may be better to have a section 'For those not familiar with Haskell' and put this under that.  The aim being making this easy for non-Haskell users to use.

fffej commented:
That's expected behaviour that at the REPL.  `sample` returns a list of tables

```
> :t (sample' (arbitrary :: Gen Table)
sample' (arbitrary :: Gen Table) :: IO [Table]
```

I hope that when you do `saveExamples` it should work as expected because (https://github.com/fffej/sql-server-gen/blob/master/src/Database/SqlServer/Generator.hs#L35) says that it should join them all together as a string and dump it to a file.  If I do

```
> x <- sample (arbitrary :: Gen Table)
> saveExamples "foo" x
> :!less "foo"
```

Then I don't see the behaviour you describe when written to a file.

Perhaps you could write a function which takes a list of entities and just turns it to a string?  

```
asScript :: Entity a => [a] -> String
```

Then we could delete the (rather daft) `saveExamples` function and just use the standard `System.IO.writeFile`?

SuperDrew commented:
Cool, I'll give that a go.

Commit:
Merge branch 'master' of https://github.com/fffej/sql-server-gen into Cabal-Order

Commit:
Use $ syntax

Commit:
Add asScript function

SuperDrew commented:
Function in, now to update read.me.

Noticed we messed up the examples with our simple options, not 100% sure how to fix this.

Commit:
Use asScript to write examples to file in Read.me.

Commit:
Add default options and use for generating entities in examples.

SuperDrew commented:
Put default in, I think this is ready for review @fffej .  Still not 100% sure this is the best way to do this.

fffej commented on src/Database/SqlServer/Generator.hs:
```diff
@@ -46,6 +46,14 @@ data GenerateOptions = GenerateOptions
   , excludeTypes :: RenderOptions
   }
 
+defaultGenerateOptions :: GenerateOptions
```
This looks great.

The only constructive feedback I can give is around the use of `undefined`.  By leaving these as `undefined`, you've got no way of doing anything nice with them.  You can't evaluate `undefined` or compare anything to is, it's a bottom type.

Would it be possible to give a default `size` and `seed`?

SuperDrew replied:
Good point.  What I really want is to only define a default for the RenderOptions and not do anything with the others. I've committed a change, what do you think? The con to my change is that you loose the ability (as far as I am aware) to specify the arguments by name.  I don't know if we care?

Commit:
Don't leave seed and size undefined.

SuperDrew commented:
Now that I deleted this branch I tested the instructions off master and `generateEntites` isn't in scope.  I can't remember why I changed the Cabal order to start in `Database`.  I tried all the examples with the old order (starting in `Generator`) and they all work :confused:.  @fffej Help

fffej commented:
Sorry, that'll teach me to read diffs and not try them out.  My bad! :-1: 

When you order the `exposed-modules` section in Cabal the first one is the one that opens.  This module imports everything (and fixes the unused warnings with the horrible hack at https://github.com/fffej/sql-server-gen/blob/master/src/Database/SqlServer/Generator.hs#L31. This should mean everything is in scope. 

I've tested the examples and they all work properly now!  I've pushed this change directly to `master` because I am a naughty man.

SuperDrew commented:
Cool, I was aware of the first one being the one it opens, but couldn't remember why I changed it.
